### PR TITLE
[ENHANCEMENT] Add a store_id property to StoreBackend

### PR DESCRIPTION
**Pull request by anthonyburdi**
Changes proposed in this pull request:
- Add a store_backend_id property to StoreBackend & Store metaclasses
- Create and persist this store_backend_id upon instantiation of a StoreBackend
- Skip validation rules for the key of a store_backend_id, do not add suffixes (store_backend_id is saved on file/blob stores as a `.ge_store_backend_id` file)
- Test in all backends (GCS testing could be better)
- Helper methods for testing

Previous Design Review notes:
- See related [design doc for in-memory configurations](https://github.com/superconductive/design/blob/main/docs/20201020_in_memory_ge.md)

Design Reviewer: @jcampbell (tagging you for review since you wrote the original design doc. Please let me know if I should reassign.)
Code Reviewer: @talagluck

### CODE REVIEW

**Comment by anthonyburdi** [comment]
I'm currently adding `store_backend_id`

File: great_expectations/data_context/store/store_backend.py
Comment (jcampbell) [inline great_expectations/data_context/store/store_backend.py]: It looks to me like this parameter is being ignored; shouldn't we set the key to this value if it's provided? It also seems to me that we should accept both the string and the UUID version of the uuid here.

**Review by jcampbell — CHANGES_REQUESTED** [review]
This looks excellent. One small nit inline. One additional request: would you add a test that shows the persistence of the store_backend_id (i.e. create two Expectations stores with the same backend *configuration* (vs. object) and verify they report the same store backend id)

File: great_expectations/data_context/store/store.py
Comment (talagluck) [inline great_expectations/data_context/store/store.py]: Per our conversation, I'm wondering if we should call this a `store_backend_id` rather than a `store_id`, since it sounds like it's possible for multiple stores to share a single store_backend.
```suggestion
        return self._store_backend.store_backend_id
```

**Review by talagluck — COMMENTED** [review]
Overall the tests make sense! Just the one naming suggestion.

File: great_expectations/data_context/store/store.py
Comment (anthonyburdi) [reply great_expectations/data_context/store/store.py]: This makes sense to me, would you mind weighing in here @jcampbell in case there is some reason I'm not thinking about?
Also to note: the way it is coded, multiple stores will report the same store_id if they share the same store_backend. I think that makes sense, but let me know if not.

File: great_expectations/data_context/store/store_backend.py
Comment (anthonyburdi) [reply great_expectations/data_context/store/store_backend.py]: A few questions after thinking about this and trying to implement:
1. I can't think of a case where we want to manually set the `store_id` rather than have it be generated for the user on instantiation or read from an existing store. We also want to make sure it is anonymized. Is it reasonable to remove this parameter? 
2. If NO to question 1: should a `store_id` set via this parameter as a `str` be validated to make sure it is a valid uuid4, or is any string sufficient in this case? I'm assuming we want to make sure it is anonymized, so probably validate.

File: great_expectations/data_context/store/store.py
Comment (jcampbell) [reply great_expectations/data_context/store/store.py]: I agree that multiple stores will report the same id. 

With respect to the concrete name: both frankly work for me, since it's kinda doing everything (*by design*) it's the id of the store backend, which then becomes the id of the store. Defer to convo you two had.

File: great_expectations/data_context/store/store_backend.py
Comment (jcampbell) [reply great_expectations/data_context/store/store_backend.py]: Good point. I think just remove the param.

Comment (anthonyburdi) [reply great_expectations/data_context/store/store_backend.py]: FYI - I'm also adding in automatic initialization of the store_backend_id in the various StoreBackend implementations when the StoreBackend is instantiated (rather than waiting for the property to be called). I think we want that feature but please let me know if that's a bad assumption!

**Review by anthonyburdi — COMMENTED** [review]
@jcampbell Is this an acceptable way to handle the non-conforming `store_backend_id` key in stores that require specific key types? https://github.com/great-expectations/great_expectations/pull/2030/commits/8f2ec5a3e0f3b5770c6d6fea13e9c9fad4d578ef

**Comment by anthonyburdi** [comment]
FYI @jcampbell - when moving the `_ = self.store_backend_id` to the end of `StoreBackend.__init__()`I get this error: `AttributeError: 'TupleFilesystemStoreBackend' object has no attribute 'forbidden_substrings'`. I think this is because the store is not yet set up fully since the `__super__` call is early in the subclass `__init__`. (this is why I moved the `_ = self.store_backend_id` to the end of the subclass init)

**Comment by anthonyburdi** [comment]
@jcampbell @talagluck - I've made quite a few changes since your reviews - could you review again before we merge into the feature branch?

**Comment by jcampbell** [comment]
> @jcampbell Is this an acceptable way to handle the non-conforming `store_backend_id` key in stores that require specific key types? [8f2ec5a](https://github.com/great-expectations/great_expectations/commit/8f2ec5a3e0f3b5770c6d6fea13e9c9fad4d578ef)

That makes sense to me, yes.

File: great_expectations/data_context/store/expectations_store.py
Comment (jcampbell) [inline great_expectations/data_context/store/expectations_store.py:102]: nit, but I believe the preferred typing solution here is `Dict` vs `dict` so that you could add in the key/value types. Not worth changing, but FYI.

File: great_expectations/data_context/store/html_site_store.py
Comment (jcampbell) [inline great_expectations/data_context/store/html_site_store.py]: another nit --> I have a pet peeve against using a contraction in code since the apostrophe is omitted. Can you rename this parameter to something like `suppress_store_backend_id`?

**Review by jcampbell — APPROVED** [review]
This looks great! I'm marking approved so I don't need to retouch, but:

1. Please note one parameter name change request
2. I didn't line-by-line reread, just looked for substantive changes. If there's something that needs a more careful code review please ensure either someone else does that or it's flagged for me.

File: great_expectations/data_context/store/expectations_store.py
Comment (anthonyburdi) [reply great_expectations/data_context/store/expectations_store.py:102]: Good call - I just checked though and since python 3.9 using `Dict` is deprecated since `builtins.dict` supports the `[]`: `https://docs.python.org/3/library/typing.html#typing.Dict`. However: "To annotate arguments it is preferred to use an abstract collection type such as Mapping." Will leave as is for now until I understand better how to use `Mapping`.

File: great_expectations/data_context/store/html_site_store.py
Comment (anthonyburdi) [reply great_expectations/data_context/store/html_site_store.py]: Done! Thanks for the suggestion, `suppress_store_backend_id` sounds much better.

File: great_expectations/data_context/store/store.py
Comment (talagluck) [inline great_expectations/data_context/store/store.py:98]: Do we want to return something or raise an error `if value == False`?

**Review by talagluck — APPROVED** [review]
Looks great! Just one piece of minor feedback.

**Pull request merged**
